### ISSUE TITLE
HTTP JSON api "Mismatching type vars and applied types ..." w/ generic template

### ISSUE THREAD

**Issue by lucianojoublanc-da** [digital-asset/daml#3488]
I am getting the following error when trying to exercise a choice on a generic template: 

```
JsonError: Mismatching type vars and applied types, expected ImmArray.ImmArraySeq(ap7) but got ImmArray.ImmArraySeq() types
```

This only happens when I exercise via the HTTP JSON API. I've tried using the Navigator, and things seem to work fine there. I've also tried replacing the generic template and  `template instance` with an instance I've written by hand, and that seems to solve the problem as well.

Steps to reproduce
-------------------------
on daml SDK v 0.13.36

  1. Navigator: create a `Callback` contract
  2. Navigator: create a `MyContinuation` contract, referencing contractId from step 1.
  3. Send the following CURL request (**replacing the contract id with that from step 2**):

```
curl -H "Authorization: Bearer eyJhbGciUzI1NiIsInR5cCI6IkpXVCJ9.eyJsZWRnZXJJZCI6Ik15TGVkZ2VyIiwiYXBwbGljYXRpb25JZCI6ImZvb2JhciIsInBhcnR5IjoiQWxpY2UifQ.4HYfzjlYr1ApUDot0a6a4zB49zS_jrwRUOCkAiPMqo0" --data '{ "templateId" : { "moduleName" : "Main", "entityName": "MyContinuation" }, "contractId" : "#114:0", "choice": "Continue", "argument": { "argsToContinuation" : { "i" : 1 } } }' http://localhost:7575/command/exercise

{"status":400,"errors":["JsonError: Mismatching type vars and applied types, expected ImmArray.ImmArraySeq(ap7) but got ImmArray.ImmArraySeq() types"]}
```

To make this work you need to start the sandbox with something like this (`daml start` won't work):

```
daml sandbox --ledgerid=MyLedger --wall-clock-time .daml/dist/*.dar & daml navigator server & (sleep 6 && daml json-api --ledger-host localhost --ledger-port 6865 --http-port 7575) &
```
You can see that the exercise works for both the scenario in the IDE, as well as exercising the choice via the Navigator.
You can also make it work by replacing `template instance ...` with the instance by hand.

Gist with *.daml model forthcoming ...

**Comment by lucianojoublanc-da** [comment]
Minimal example to reproduce:

https://gist.github.com/lucianojoublanc-da/70d8e676b22c1b5e0fe4647747a03c0f

**Comment by hurryabit** [comment]
My first guess is that the JSON API is written under the assumption that the types of choice arguments have to be of a specific form (most likely record types without any type arguments). However, the DAML-LF spec says they can be arbitrary serializable ground types, i.e., serializable types without free type variables.

Choices on generic templates are the first choices `damlc` produces for which this assumption does not hold.

**Comment by cocreature** [comment]
Probably related to #3390

**Comment by hurryabit** [comment]
Most likely a duplicate of #3390.

**Comment by leo-da** [comment]
@cocreature @hurryabit 
how do you expect choice arguments on a generic template?

**Comment by cocreature** [comment]
@leo-da On a generic template, the choice argument will be a record with type parameters.

**Comment by leo-da** [comment]
@cocreature any references to the doc/spec?

**Comment by cocreature** [comment]
The DAML-LF spec should be sufficient https://github.com/digital-asset/daml/blob/master/daml-lf/spec/daml-lf-1.rst. From the DAML-LF perspective there is nothing special about the choice record type.

**Comment by cocreature** [comment]
Closing since generic templates have been removed. The general issue still exists but is covered by #3390

**Comment by leo-da** [comment]
The root problem is the JSON to LF Value decoder. Here is how we currently create an `iface.TypeCon`, note that we pass empty `typArgs`:
```
val typeCon = Model.DamlLfTypeCon(Model.DamlLfTypeConName(id), ImmArraySeq())
```

https://github.com/digital-asset/daml/blob/a6f3dc309f73d6d3cefff488c09180a75fdd6f03/ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala#L285-L295

`InterfaceReader` actually needs to change to allow looking up `iface.TypeCon`s with all `typArgs`. Need something like:
```
com.digitalasset.daml.lf.data.Ref.Identifier => com.digitalasset.daml.lf.iface.TypeCon
```

**Comment by leo-da** [comment]
This particular issue has been addressed as part of #3882. We use a different call:
https://github.com/digital-asset/daml/blob/a6f3dc309f73d6d3cefff488c09180a75fdd6f03/ledger-service/lf-value-json/src/main/scala/com/digitalasset/daml/lf/value/json/ApiCodecCompressed.scala#L255-L258

which takes `com.digitalasset.daml.lf.iface.Type`.

However when we introduce generic templates back, we will need `InterfaceReader` to provide type details about template record type arguments.
